zio-http/shared/src/main/scala/zio/http/ZClient.scala

```diff
@@ -19,25 +19,27 @@ package zio.http
 import java.net.{InetSocketAddress, URI}
 
 import zio._
+import zio.stacktracer.TracingImplicits.disableAutoTrace
+
+import zio.stream.ZStream
 
 import zio.http.URL.Location
 import zio.http.internal._
-//import zio.http.netty.client._
 
-final case class ZClient[-Env, -In, +Err, +Out](
+final case class ZClient[-Env, S >: Scope, -In, +Err, +Out](
```

Comment:
1. In general, I am not a fan of constraining type parameters at the declaration site, unless absolutely strictly required. The constraints add a lot of work for no real benefit and also interfere with evolution.
2. Because this `S` is added to each request response (covariant), and there it appears in contravariant position, we should be able to make the parameter contravariant, unless I'm missing something.

This would change the signature like so:

```suggestion
final case class ZClient[-Env, -S, -In, +Err, +Out](
```

In which case I might rename:

```suggestion
final case class ZClient[-Env, -ReqEnv, -In, +Err, +Out](
```

"The env added to requests."

Or something similar (`AddEnv`).

Reply:
> I am not a fan of constraining type parameters at the declaration site

Main reason that I went with the constraining type is that removing it doesn't affect binary compatibility, so I thought perhaps it's better to be more strict and remove it if needed. But I agree I don't think that it adds much benefit so I'll remove it 

> we should be able to make the parameter contravariant, unless I'm missing something

The problem with making it contravariant is that then this becomes possible:

```scala
  val batchedClient: ZClient[Any, Any, Body, Throwable, Response] = ???
  val streamingClient: ZClient[Any, Scope, Body, Throwable, Response] = batchedClient
```

This is a problem because once we've eliminated the `Scope` the client cannot stream thereafter. The only way to enforce it AFAICT is via an invariant type param